File: leetcode questions/0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;
        return searchRange(nums, target, start, end);
    }

    public int[] searchRange(int[] nums, int target, int start, int end) {
        int[] array = {-1, -1};
        if (start > end) return array;
        int pivot = (start + end) / 2;

        if (nums[pivot] == target) {
            int left = search(nums, target, start, pivot - 1, true);
            if (left == -1) left = pivot;
            int right = search(nums, target, pivot + 1, end, false);
            if (right == -1) right = pivot;
            array[0] = left;
            array[1] = right;
            return array;
        } else if (nums[pivot] < target) {
            return searchRange(nums, target, pivot + 1, end);
        }
        return searchRange(nums, target, start, pivot - 1);
    }

    public int search(int[] nums, int target, int start, int end, boolean isLeft) {
        if (start > end) return -1;
        int pivot = (start + end) / 2;
        if (isLeft) {
            if (nums[pivot] == target) {
                if (pivot == 0 || nums[pivot - 1] < target) {
                    return pivot;
                } else {
                    end = pivot - 1;
                }
            } else {
                start = pivot + 1;
            }
        } else {
            if (nums[pivot] == target) {
                if (pivot == nums.length - 1 || nums[pivot + 1] > target) {
                    return pivot;
                } else {
                    start = pivot + 1;
                }
            } else {
                end = pivot - 1;
            }
        }
        return search(nums, target, start, end, isLeft);
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
    private:
    int firstOcc(vector<int>& nums, int target){
        int low=0;
        int n=nums.size();
        int high=n-1;

        int first=-1;
        while(low<=high){
            int mid=(low+high)/2;
            if(nums[mid]==target){
                first=mid;
                high=mid-1;
            }
            else if(nums[mid]<target){
                low=mid+1;
            }else{
                high=mid-1;
            }
        }
        return first;
    }
    int lastOcc(vector<int>& nums, int target){
        int low=0;
        int n=nums.size();
        int high=n-1;

        int last=-1;
        while(low<=high){
            int mid=(low+high)/2;
            if(nums[mid]==target){
                last=mid;
                low=mid+1;
                
            }
            else if(nums[mid]<target){
                low=mid+1;
            }else{
                high=mid-1;
            }
        }
        return last;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
       int first=firstOcc(nums,target);
       if(first==-1) return {-1,-1};
       int last=lastOcc(nums,target);
       return {first,last};
    }
};